File: sdk/python/costorah/_logging.py
"""
Structured logging with mandatory secret redaction — the SDK must never
log an API key, a prompt, a model response, or other sensitive metadata,
even at DEBUG level. This is enforced here rather than trusted to call
sites, mirroring the Monitoring Agent's `logging_setup.py`
(`monitoring-agent/costorah_agent/logging_setup.py`, EP-17) so the whole
COSTORAH ecosystem redacts consistently — this is a parallel
implementation, not a shared import (the SDK does not depend on the
Monitoring Agent package).
"""

from __future__ import annotations

import logging
import re
from typing import Any

_REDACTED_KEY_PATTERNS = (
    "api_key",
    "apikey",
    "authorization",
    "password",
    "secret",
    "token",
    "prompt",
    "completion",
    "response_body",
    "user_prompt",
    "model_response",
)

_BEARER_TOKEN_RE = re.compile(r"costorah_live_[A-Za-z0-9_-]+")


def _redact_string(value: str) -> str:
    return _BEARER_TOKEN_RE.sub("costorah_live_***REDACTED***", value)


def redact(value: Any) -> Any:
    """Recursively redact known-sensitive keys and any embedded
    costorah_live_ token substring, regardless of which field it's in."""
    if isinstance(value, str):
        return _redact_string(value)
    if isinstance(value, dict):
        redacted: dict[Any, Any] = {}
        for key, val in value.items():
            if isinstance(key, str) and any(
                pattern in key.lower() for pattern in _REDACTED_KEY_PATTERNS
            ):
                redacted[key] = "***REDACTED***"
            else:
                redacted[key] = redact(val)
        return redacted
    if isinstance(value, (list, tuple)):
        return type(value)(redact(item) for item in value)
    return value


class _RedactingFilter(logging.Filter):
    def filter(self, record: logging.LogRecord) -> bool:
        record.msg = redact(record.msg) if isinstance(record.msg, str) else record.msg
        if record.args:
            if isinstance(record.args, dict):
                record.args = redact(record.args)
            else:
                record.args = tuple(redact(a) for a in record.args)
        return True


def get_logger(name: str = "costorah") -> logging.Logger:
    """Return a logger with the redaction filter attached. Safe to call
    repeatedly — the filter is only attached once per logger instance."""
    logger = logging.getLogger(name)
    if not any(isinstance(f, _RedactingFilter) for f in logger.filters):
        logger.addFilter(_RedactingFilter())
    return logger

File: sdk/javascript/src/logging.ts
/**
 * Structured logging with mandatory secret redaction — the SDK must never
 * log an API key, a prompt, a model response, or other sensitive
 * metadata, even at debug level. Mirrors the Python SDK's `_logging.py`
 * and the Monitoring Agent's `logging_setup.py` (EP-17) so the whole
 * COSTORAH ecosystem redacts consistently.
 */

const REDACTED_KEY_PATTERNS = [
  "apikey",
  "api_key",
  "authorization",
  "password",
  "secret",
  "token",
  "prompt",
  "completion",
  "responsebody",
  "response_body",
  "userprompt",
  "user_prompt",
  "modelresponse",
  "model_response",
];

const BEARER_TOKEN_RE = /costorah_live_[A-Za-z0-9_-]+/g;

function redactString(value: string): string {
  return value.replace(BEARER_TOKEN_RE, "costorah_live_***REDACTED***");
}

/** Recursively redact known-sensitive keys and any embedded
 * costorah_live_ token substring, regardless of which field it's in. */
export function redact(value: unknown): unknown {
  if (typeof value === "string") {
    return redactString(value);
  }
  if (Array.isArray(value)) {
    return value.map(redact);
  }
  if (value !== null && typeof value === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, val] of Object.entries(value)) {
      const lowered = key.toLowerCase();
      if (REDACTED_KEY_PATTERNS.some((pattern) => lowered.includes(pattern))) {
        result[key] = "***REDACTED***";
      } else {
        result[key] = redact(val);
      }
    }
    return result;
  }
  return value;
}

export type LogLevel = "debug" | "info" | "warn" | "error";

export interface Logger {
  debug(message: string, fields?: Record<string, unknown>): void;
  info(message: string, fields?: Record<string, unknown>): void;
  warn(message: string, fields?: Record<string, unknown>): void;
  error(message: string, fields?: Record<string, unknown>): void;
}

const LEVEL_ORDER: Record<LogLevel, number> = { debug: 0, info: 1, warn: 2, error: 3 };

/** A minimal structured console logger with redaction baked in. Callers
 * may supply their own `Logger` implementation (e.g. to route into pino
 * or winston) as long as it redacts consistently — the SDK always calls
 * `redact()` before handing fields to whichever logger is configured, so
 * even a caller-supplied logger receives already-redacted data. */
export function createConsoleLogger(level: LogLevel = "info"): Logger {
  const threshold = LEVEL_ORDER[level];
  const emit = (lvl: LogLevel, message: string, fields?: Record<string, unknown>): void => {
    if (LEVEL_ORDER[lvl] < threshold) return;
    const safeFields = fields ? redact(fields) : undefined;
    const line = `[costorah] ${lvl.toUpperCase()} ${message}`;
    // eslint-disable-next-line no-console
    const consoleMethod = lvl === "error" ? console.error : lvl === "warn" ? console.warn : console.log;
    if (safeFields) {
      consoleMethod(line, safeFields);
    } else {
      consoleMethod(line);
    }
  };

  return {
    debug: (message, fields) => emit("debug", message, fields),
    info: (message, fields) => emit("info", message, fields),
    warn: (message, fields) => emit("warn", message, fields),
    error: (message, fields) => emit("error", message, fields),
  };
}
